### PR TITLE
RasterHeader for storing raster meta data

### PR DESCRIPTION
**Pull request by rinkk**
this is the first follow-up on #1004 (and should be merged *after* #1004)
it introduces a tiny struct RasterHeader to story raster meta data and uses this throughout ogs where raster data is handled. it basically replaces 6 parameters in method-calls with one. it also fixes an index bug in Raster to VtkImageData conversion.

### CODE REVIEW

File: Applications/DataExplorer/VtkVis/VtkRaster.cpp [hunk 1]
```diff
@@ -94,10 +86,10 @@ vtkImageImport* VtkRaster::loadImageFromArray(double const*const data_array, dou
...
 
 	vtkImageImport* image = vtkImageImport::New();
-		image->SetDataSpacing(delta, delta, delta);
-		image->SetDataOrigin(x0+(delta/2.0), y0+(delta/2.0), 0);	// translate whole mesh by half a pixel in x and y
-		image->SetWholeExtent(0, width-1, 0, height-1, 0, 0);
-		image->SetDataExtent(0, width-1, 0, height-1-1, 0, 0);
+		image->SetDataSpacing(header.cell_size, header.cell_size, header.cell_size);
+		image->SetDataOrigin(header.origin[0]+(header.cell_size/2.0), header.origin[1]+(header.cell_size/2.0), 0);	// translate whole mesh by half a pixel in x and y
+		image->SetWholeExtent(0, header.n_rows-1, 0, header.n_cols-1, 0, 0);
+		image->SetDataExtent(0, header.n_rows-1, 0, header.n_cols-1-1, 0, 0);
```
Comment (wenqing) [inline Applications/DataExplorer/VtkVis/VtkRaster.cpp]: header.n_cols-1-1 -->header.n_cols - 2

File: Applications/DataExplorer/VtkVis/VtkVisPipelineView.cpp [hunk 2]
```diff
@@ -196,13 +196,15 @@ void VtkVisPipelineView::addPipelineFilterItem()
...
+	MeshFromRasterDialog dlg;
+	if (dlg.exec() != QDialog::Accepted)
+		return;
+
+	MeshLib::MeshElemType t (dlg.getElementSelection());
+	constructMeshFromImage(dlg.getNewMeshName(), dlg.getElementSelection(), dlg.getIntensitySelection());
 }
 
-void VtkVisPipelineView::constructMeshFromImage(QString msh_name, MeshLib::MeshElemType element_type, MeshLib::UseIntensityAs intensity_type)
+void VtkVisPipelineView::constructMeshFromImage(std::string msh_name, MeshLib::MeshElemType element_type, MeshLib::UseIntensityAs intensity_type)
```
Comment (TomFischer) [inline Applications/DataExplorer/VtkVis/VtkVisPipelineView.cpp]: Why not `std::string const & msh_name`?

File: GeoLib/Raster.h [hunk 3]
```diff
@@ -58,32 +61,35 @@ class Raster {
...
 	 */
-	double getRasterPixelSize() const { return _cell_size; }
+	double getRasterPixelSize() const { return _header.cell_size; }
 
 	/**
 	 * get the origin of lower left raster cell
 	 * @return the origin of the raster
 	 */
-	GeoLib::Point const& getOrigin() const;
+	GeoLib::Point const getOrigin() const;
```
Comment (TomFischer) [inline GeoLib/Raster.h]: Why not return a `const &` here?

**Comment by TomFischer** [comment]
Good idea! :+1:

**Comment by endJunction** [comment]
The changes in the last commit 'rebase' a785ff8 revert some changes in the same pr. Is it intended?

Also check for trailing whitespaces.... It would be nice if the code base stays in consistent format.

File: Applications/DataExplorer/DataView/NetCdfConfigureDialog.cpp [hunk 4]
```diff
@@ -352,13 +354,14 @@ void NetCdfConfigureDialog::createDataObject()
 
 	double origin_x = (originLon < lastLon) ? originLon : lastLon;
 	double origin_y = (originLat < lastLat) ? originLat : lastLat;
-	double originNetCdf[3] = {origin_x, origin_y, 0};
+	MathLib::Point3d origin(std::array<double,3>{{origin_x, origin_y, 0}});
```
Comment (endJunction) [inline Applications/DataExplorer/DataView/NetCdfConfigureDialog.cpp]: @TomFischer Would be nice if the Point3d could be directly initialized from an initializer list.

File: Applications/DataExplorer/VtkVis/VtkRaster.cpp [hunk 5]
```diff
@@ -54,15 +54,8 @@ vtkImageAlgorithm* VtkRaster::loadImage(const std::string &fileName,
...
-		x0 = raster->getOrigin()[0];
-		y0 = raster->getOrigin()[1];
-		delta = raster->getRasterPixelSize();
-		double const*const data (raster->begin());
-		return VtkRaster::loadImageFromArray(data, x0, y0,
-						raster->getNCols(), raster->getNRows(), delta,
-						raster->getNoDataValue());
-	}
+	if (raster)
+		return VtkRaster::loadImageFromArray(raster->begin(), raster->getHeader());
```
Comment (endJunction) [inline Applications/DataExplorer/VtkVis/VtkRaster.cpp]: Would you please explain how the `raster` is deleted after this return? It looks like a memory leak to me.

File: FileIO/AsciiRasterInterface.cpp [hunk 6]
```diff
@@ -43,24 +43,21 @@ GeoLib::Raster* AsciiRasterInterface::getRasterFromASCFile(std::string const& fn
...
 
 	// header information
-	std::size_t n_cols(0), n_rows(0);
-	double xllcorner(0.0), yllcorner(0.0), cell_size(0.0), no_data_val(-9999);
-
-	if (readASCHeader(in, n_cols, n_rows, xllcorner, yllcorner, cell_size, no_data_val)) {
-		double* values = new double[n_cols*n_rows];
+	GeoLib::RasterHeader header;
+	if (readASCHeader(in, header)) {
+		double* values = new double[header.n_cols*header.n_rows];
```
Comment (endJunction) [inline FileIO/AsciiRasterInterface.cpp]: It is better to use unique_ptr here.

File: MeshLib/MeshGenerators/RasterToMesh.h [hunk 7]
```diff
@@ -0,0 +1,125 @@
+/**
+ * \file
+ * \author Karsten Rink
+ * \date   2011-08-23
+ * \brief  Definition of the VtkMeshConverter class.
```
Comment (endJunction) [inline MeshLib/MeshGenerators/RasterToMesh.h]: Please syncronize the documentation with the file content.

**Comment by endJunction** [comment]
There are two new files, RasterToMesh.{cpp,h} which look similar to ConvertRasterToMesh.cpp. Is it just a copy, or are there other changes?

**Comment by rinkk** [comment]
ConvertRasterToMesh was a partial copy of VtkMeshConverter. Then VtkMeshConverter was changed which made the copy unnecessary. With the above changes I deleted ConvertRasterToMesh and split VtkMeshConverter into a mesh conversion part and an raster conversion part.

File: FileIO/AsciiRasterInterface.cpp (hunk 6 above)
Comment (rinkk) [inline FileIO/AsciiRasterInterface.cpp]: We can do this later, it has nothing to do with this PR

File: Applications/DataExplorer/VtkVis/VtkRaster.cpp (hunk 5 above)
Comment (rinkk) [inline Applications/DataExplorer/VtkVis/VtkRaster.cpp]: not part of this PR, i just did the syntactic changes for using RasterHeader.
i suppose i could change raster to a unique_ptr to solve the problem.

**Comment by rinkk** [comment]
concerning the rebase: there was a merge conflict and I added the code from the current master at that position.

**Comment by endJunction** [comment]
Looks good. :fast_forward: 

Regarding the 'rebase' commit, it only seemed strange to me that there were changes in the PR which were then removed while resolving a conflict. That could be an error.

Please rebase on current master and merge when tests are passing.
